### ISSUE TITLE
maple.c - Expanding JVS Functionality (add Coin support, etc.)

### ISSUE THREAD

**Issue by Softdrink117** [DragonMinded/libnaomi#6]
I'm working on expanding the JVS functionality (EG: to add Coin input and decrement support), and ran into a few issues that I want to discuss before I continue, since I am proposing several significant changes (though I believe none of them would break existing APIs):

1. The `__outstanding_request` flags can cause unexpected behavior when mixed with other JVS functions. Once `maple_request_jvs_buttons` has been called at least once, that function will cease to work if another JVS message is sent without unsetting (or re-requesting) the `__outstanding_request` flags. My current workaround for this is to simply retry any outstanding request, if one exists, as part of other JVS functions, EG:
```
if(__outstanding_request || __outstanding_request_addr != 0)
    {
        __outstanding_request = __maple_request_jvs_send_buttons_packet(__outstanding_request_addr, 1);
    }
```
But I'm wondering if you have any thoughts about a better way to handle this process, since this is rather wasteful (the original `__outstanding_request` is not used, burning maple bus cycles and time).

2. The existing `_maple_request_send_jvs()` function limits the usability of JVS features (such as concatenated multi-command Requests) because of the small send buffer size. I'd like to propose expanding the `subcommand` buffer size, and passing the payload length to `_maple_swap_data` (right now it is hardcoded as 3 bytes, regardless of the `len` passed to `_maple_request_send_jvs`). Is there a specific reason why these limits are in place, or can they be larger?

3. The current implementation does not expose lower-level JVS functionality outside of what libnaomi offers in `maple.h`. While this is probably more than adequate for many homebrew use cases, there are quite a few additional JVS functions (coins, alternative input devices, controllable cabinet hardware, etc.) that are not currently accessible. I'd propose offering lower-level access to those who desire it, by exposing `_maple_request_send_jvs`, `_maple_request_recv_jvs`, and some of the related data types and helper functions (`jvs_status_t`, etc.). It's technically possible to implement some extensions outside of libnaomi using these functions right now using `extern`, but being able to include them in the `maple.h` header would be more resilient to future implementation changes in libnaomi.

**Comment by DragonMinded** [comment]
Honestly, the whole maple underlying subsystem should probably be its own thread with a state system and such. That's really what all of the button handling functions are trying to emulate with "__outstanding_request". As it stands, it makes a LOT of assumptions about what traffic is going on the MAPLE bus like you've pointed out. I wrote most of the maple code before there was a working threading system or a kernel, so it was written the way it is to get off the ground and working. It was also not written extensibly, simply written to get things off the ground and usable. Some things like coins and counters should probably be part of the main MAPLE high-level interface, but also there should be a way to do lower-level JVS things directly.

As long as the external API is preserved and it works as expected from the examples, basically any and all of the MAPLE internals can be redone in a way that facilitates better extension. I think that's entirely possible given the way things are laid out, and libnaomi makes NO promises to the stability of non-public functions. The MAPLE subsystem has its own init and free which is hooked into the core startup/shutdown routines, so it should be safe to spawn off a thread if needed. Or maybe, you have ideas on how to handle this cleaner without needing a daemon for talking to MAPLE? Pretty sure official games abstract with a thread.

**Comment by Softdrink117** [comment]
Ah, okay, that gives a lot more context.

From my end, I'm honestly a bit overwhelmed at this point in time trying to understand how the MAPLE system works, and was hoping to add only a little bit of layering on top to get the features I want working. It seems like I might've stumbled into a much bigger project than I intended, heh.

I can absolutely take a stab at some of this more substantial rework, but my own knowledge of how the MAPLE system operates is pretty limited right now and I'll need to do a lot more digging first.

Also, in the process of trying to test some of my own extensions, I seem to have entered an _extremely spooky_ state when testing on hardware where no libnaomi-based inputs work (including previously known working sample binaries), but official games work fine. So that's... _**fun**_...

**Comment by Softdrink117** [comment]
OK, the spooky issue appears to be a problem with my IO that _just happened to coincide_ with my tests with maple. Which is a huge relief, what a crazy coincidence.

Repaired my IO (firmware reflash) and it's working again; then I tested with an official Sega IO and that's also working.

So far, I've managed to validate that most of my proposed changes are working; others need a bit more polish. I'll dig into this a bit more and see where it goes.
